android/src/main/java/com/matejdr/admanager/BannerAdView.java

```diff
@@ -118,7 +133,114 @@ public void onAdClosed() {
             }
 
         });
+
+        updateLayout();
+
         this.addView(this.adView);
+
+        updateLayout();
+    }
+
+    private class MeasureAndLayoutRunnable implements Runnable {
+        @Override
+        public void run() {
+            updateLayout();
+        }
+    }
+
+    private boolean isFluid() {
+        if (this.adSize == null) {
+            return false;
+        }
+
+        return this.adSize.equals(AdSize.FLUID);
+    }
+
+    @Override
+    public void requestLayout() {
+        super.requestLayout();
+
+        if (isFluid()) {
+            post(new MeasureAndLayoutRunnable());
+        }
+    }
+
+    @Override
+    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
+        super.onLayout(changed, left, top, right, bottom);
+
+        if (isFluid()) {
+            post(new MeasureAndLayoutRunnable());
+        }
+    }
+
+    private static void measureAndLayout(View view, int width, int height) {
+        int left = 0;
+        int top = 0;
+
+        view.measure(width, height);
+        view.layout(left, top, left + width, top + height);
+        view.requestLayout();
+        view.invalidate();
+        view.forceLayout();
+    }
+
+    int cachedWidth = 0;
+    int cachedHeight = 0;
+
+    private void updateLayout() {
+        try {
+            if (!isFluid()) {
+                return;
+            }
+
+            if (adView == null) {
+                return;
+            }
+
+            View parent = (View) adView.getParent();
+
+            if (parent == null) {
+                return;
+            }
+
+            int width = parent.getWidth();
+            int height = parent.getHeight();
+
+            if (cachedWidth == width && cachedHeight == height) {
+                return;
+            }
+
+            cachedWidth = width;
+            cachedHeight = height;
+
+            measureAndLayout(adView, width, height);
+
+            ViewGroup child = (ViewGroup) adView.getChildAt(0);
+
+            if (child != null) {
+                measureAndLayout(child, width, height);
+
+                ViewGroup webView = (ViewGroup) child.getChildAt(0);
+
+                if (webView != null) {
+                    measureAndLayout(webView, width, height);
+
+                    ViewGroup internalChild = (ViewGroup) webView.getChildAt(0);
+
+                    if (internalChild != null) {
+                        measureAndLayout(internalChild, width, height);
+
+                        ViewGroup leafNode = (ViewGroup) internalChild.getChildAt(0);
+
+                        if (leafNode != null) {
+                            measureAndLayout(leafNode, width, height);
+                        }
+                    }
+                }
+            }
```

Comment:
can you leave a comment that explains why you are iterating over four different views?

Reply:
Basically, this is the root fix, as Google Ads can't into layout, we have to hack around and fix their broken layout implementation on Android 🤷

Reply:
don't tell me - tell the future developer working on this code ;)

Reply:
Hopefully, nobody will need that 😉 I think git blame and the PR link will bring the full context.

Reply:
I still think a comment might be a good idea. Come on, it's not killing you ;) this is code that could (ideally) end up back in the main repo. This is open source. Please write a comment describing the need for this stacked code.

Reply:
I left a comment in https://github.com/Freddixx/react-native-ad-manager/pull/4/commits/14cad93ebd3b2843390a7f73d7630eb87a08445b, feel free to suggest improvements if needed :)